smart_open/tests/test_smart_open.py

```diff
@@ -200,93 +202,158 @@ def test_http_pass(self):
-            test_file = infile.name
+        self._test_compressed_http(".xz", False)
 
-        with smart_open.smart_open(test_file, 'wb') as outfile:
-            outfile.write(test_string)
+    @unittest.skipIf(six.PY2, 'gzip support for Py2 is not implemented yet')
+    def test_http_gz_query(self):
+        """Can open gzip via http with a query appended to URI?"""
+        self._test_compressed_http(".gz", True)
 
-        with open(test_file, 'rb') as infile:
-            compressed_data = infile.read()
+    def test_http_bz2_query(self):
+        """Can open bzip2 via http with a query appended to URI?"""
+        self._test_compressed_http(".bz2", True)
 
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
+    def test_http_xz_query(self):
+        """Can open xz via http with a query appended to URI?"""
+        self._test_compressed_http(".xz", True)
 
-        responses.add(responses.GET, "http://127.0.0.1/data.xz", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.xz")
 
-        # decompress the xz and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
+class SmartOpenFileObjTest(unittest.TestCase):
+    """
+    Test passing raw file objects.
+    """
+    text = 'Hello, world!'
 
+    def patch_string_io(func):
+        if not PY2:
+            # Python 3 has already got us covered
+            return func
+
+        def enter(self):
+            return self
+
+        def exit(self, exc_type, exc_val, exc_tb):
+            self.close()
+
+        @functools.wraps(func)
+        def wrapped_patch_string_io(self):
+            six.StringIO.__enter__ = enter
+            six.StringIO.__exit__ = exit
+            try:
+                return func(self)
+            finally:
+                del six.StringIO.__enter__
+                del six.StringIO.__exit__
+
+        return wrapped_patch_string_io
+
+    @patch_string_io
+    def test_read_bytes(self):
+        """Can we read bytes from a byte stream?"""
+        buffer = six.BytesIO(self.text.encode())
+        with smart_open.smart_open(buffer, 'rb') as sf:
+            data = sf.read()
+        self.assertEqual(data, self.text.encode())
+
+    @patch_string_io
+    def test_write_bytes(self):
+        """Can we write bytes to a byte stream?"""
+        buffer = six.BytesIO()
+        with smart_open.smart_open(buffer, 'wb') as sf:
+            sf.write(self.text.encode())
+            self.assertEqual(buffer.getvalue(), self.text.encode())
+
+    @unittest.skipIf(PY2, "Python 2 does not make a difference between str and bytes")
+    @patch_string_io
+    def test_read_str(self):
+        """Can we read strings from a text stream?"""
+        buffer = six.StringIO(self.text)
+        with smart_open.smart_open(buffer, 'r') as sf:
+            self.assertRaises(TypeError, sf.read)  # we expect binary mode
```

Comment:
Sorry, I don't understand: what's the problem here?

- There's a text buffer
- You've opened it in text mode for reading
- You're trying to read from it

Why do "we expect binary mode"? Why is this test supposed to fail?

Reply:
https://github.com/RaRe-Technologies/smart_open/pull/263#discussion_r260144512 binary and text buffers behave differently on Python 3

Reply:
OK, as above then, let's have a more informative test name docstring.